test.py

```diff
@@ -1,124 +1,527 @@
+    def setUp(self):
+        self.handler = PayloadHandler()
+
+
+    def test_handle_payload(self):
+        with self.assertRaises(NotImplementedError):
+            self.handler.handle_payload("payload")
+
+
+class TestTravisPayloadHandler(unittest.TestCase):    
+    def setUp(self):
+        class TravisDouble():
+            def get_build(self, build_id):
+                return 1
+
+
+            def get_log(self, build_data):
+                return open('resources/multi-line-comment.log').read()
+
+
+            def get_pull_request_number(self, build_data):
+                return 1
+
+
+        class ErrorParserDouble():
+            path_key = ServoErrorLogParser.path_key
+            body_key = ServoErrorLogParser.body_key
+            position_key = ServoErrorLogParser.position_key
+
+            def parse_log(self, log):
+                return error_sample
+
+
+        travis_dbl = TravisDouble()
+        error_parser_dbl = ErrorParserDouble()
+        self.github = GithubApiProvider("jdm", "a453b3923e893f0383cd2893f...", "servo", "servo")
+        self.github.post_review_comment = Mock()
+        self.github.get_review_comments = Mock(return_value=json.loads(open('resources/review_comments.json').read()))
+
+        self.payload_handler = TravisPayloadHandler(self.github, travis_dbl, error_parser_dbl)
+        
+
+    def test_handle_payload(self):
+        payload = json.loads(open('resources/test_travis_payload.json').read())
+        self.payload_handler.handle_payload(payload)
+        err_msg = TravisPayloadHandler.msg_template
+
+        self.github.post_review_comment.assert_called_with(1, "9b6313fd5ab92de5a3fd9f13f8421a929b2a8ef6", err_msg.format(error_sample[3]['path'], error_sample[3]['position'], error_sample[3]['body']), error_sample[3]['path'], error_sample[3]['position'])
+
+class TestGithubPayloadHandler(unittest.TestCase):
+    def setUp(self):
+        self.github_user = "jdm"
+        self.github = GithubApiProvider(self.github_user, "a453b3923e893f0383cd2893f...", "servo", "servo")
+        self.github.remove_label = Mock()
+        self.github.add_label = Mock()
+
+
+    def test_handle_payload(self):
+        pl_handler = GithubPayloadHandler(None)
+        pl_handler.new_pr = Mock()
+        pl_handler.update_pr = Mock()
+        pl_handler.new_comment = Mock()
+
+        payload = {"action":"created", "issue":{"number":1}}
+        pl_handler.handle_payload(payload)
+        pl_handler.new_comment.assert_called_with("1", payload)
+
+        payload = {"action":"opened", "number":"1"}
+        pl_handler.handle_payload(payload)
+        pl_handler.new_pr.assert_called_with("1", payload)
+
+        payload = {"action":"synchronize", "number":"1"}
+        pl_handler.handle_payload(payload)
+        pl_handler.new_pr.update_pr("1", payload)
+
+        # should not break
+        payload = {"action":"other", "number":"1"}
+        pl_handler.handle_payload(payload)
+
+
+    def test_manage_pr_state(self):
```

Comment:
This test has some nice properties, in that it does end up verifying many of the properties that we care about, but I still am not wild about how it's directly coupled to the implementation of payload handler, rather than just the expected outcome.